src/Workspaces/Remote/Core/Storage/RemotePersistentStorageLocationService.cs

```diff
@@ -1,63 +1,19 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System;
-using System.Collections.Generic;
 using System.Composition;
-using System.IO;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
 
 namespace Microsoft.CodeAnalysis.Remote.Storage
 {
+    // Have to re-export this in remote layer for it to be picked up.
     [ExportWorkspaceService(typeof(IPersistentStorageLocationService), layer: WorkspaceKind.RemoteWorkspace), Shared]
-    internal class RemotePersistentStorageLocationService : IPersistentStorageLocationService
+    internal class RemoteWorkspacePersistentStorageLocationService : DefaultPersistentStorageLocationService
     {
-        private readonly object _gate = new object();
-        private readonly Dictionary<SolutionId, string> _idToStorageLocation = new Dictionary<SolutionId, string>();
-
-        [ImportingConstructor]
-        public RemotePersistentStorageLocationService()
-        {
-        }
-
-        public event EventHandler<PersistentStorageLocationChangingEventArgs> StorageLocationChanging;
-
-        public string TryGetStorageLocation(SolutionId solutionId)
-        {
-            _idToStorageLocation.TryGetValue(solutionId, out var result);
-            return result;
-        }
-
-        public bool IsSupported(Workspace workspace)
-        {
-            lock (_gate)
-            {
-                return _idToStorageLocation.ContainsKey(workspace.CurrentSolution.Id);
-            }
-        }
-
-        public void UpdateStorageLocation(SolutionId id, string storageLocation)
-        {
-            lock (_gate)
-            {
-                // We can get null when the solution has no corresponding file location
-                // in the host process.  This is not abnormal and can come around for
-                // many reasons.  In that case, we simply do not store a storage location
-                // for this solution, indicating to all remote consumers that persistent
-                // storage is not available for this solution.
-                if (storageLocation == null)
-                {
-                    _idToStorageLocation.Remove(id);
-                }
-                else
-                {
-                    // Store the esent database in a different location for the out of proc server.
-                    storageLocation = Path.Combine(storageLocation, "Server");
-                    _idToStorageLocation[id] = storageLocation;
-                }
-            }
+    }
```

Comment:
❓ Why is the remote workspace not picking up `DefaultPersistentStorageLocationService` (which is exported without a layer restriction)?

Reply:
Unfortunately, the remote workspace is *truly weird*.  It has it's own entirely separate MEF stack with very specific importing rules.  I really don't like it as it means that the normal workspace MEF exports don't work for it and so many things need to be special cased like this.

I would *love* to get rid of this, but that's a larger piece of work that i don't feel comfortable doing just now.  @jasonmalinowski and you would likely be great for this as i think your mef skills are def better than mine as well.

Reply:
> with very specific importing rules

Can you point me to where this is implemented?

Reply:
Sure:

https://github.com/dotnet/roslyn/blob/5882a87604eac9bd4fead4d1bf0249350d60400e/src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs#L29-L30
https://github.com/dotnet/roslyn/blob/5882a87604eac9bd4fead4d1bf0249350d60400e/src/Workspaces/Remote/Core/Services/TemporaryWorkspace.cs#L16-L17
https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Remote/Core/Services/RoslynServices.cs

No clue what the rhyme/reason is for how mef is done this way for OOP.

Reply:
OK, so it turns out that if you remove these two definitions, OOP _will_ pick up `DefaultPersistentStorageLocationService`. This occurs because this check fails:

https://github.com/dotnet/roslyn/blob/5bcd0e0834791fe659e5798a352be3d429b7df21/src/Workspaces/Core/Portable/Workspace/Host/Mef/MefWorkspaceServices.cs#L76-L80

Then this check succeeds:

https://github.com/dotnet/roslyn/blob/5bcd0e0834791fe659e5798a352be3d429b7df21/src/Workspaces/Core/Portable/Workspace/Host/Mef/MefWorkspaceServices.cs#L100-L104

We would only need to provide custom implementation(s) if we are changing some behavior from the base type, which in this case we clearly are not.

Reply:
There are two things here to investigate:

1. This pull request claims to fix #39653, but neither of the implementations in this file override the `IsSupported` method from the base. Are we sure that the issue is fixed by this pull request?
2. Can we update the `DefaultPersistentStorageLocationService.IsSupported` method to leverage a helper similar to `RemoteFeatureOptions.ShouldComputeIndex`, and then delete the two implementations in this file? Note that we can't use `ShouldComputeIndex` directly because we need to use the database in both in-proc and OOP at the same time (though with different locations on disk).